File: frontend/src/types/job.ts
/**
 * Pipeline One - Job type definitions
 */

export type JobStatus = 'pending' | 'running' | 'paused' | 'completed' | 'failed' | 'cancelled';

export type PipelineStage = 'extraction' | 'segmentation' | 'reconstruction' | 'tracking';

export const ALL_PIPELINE_STAGES: PipelineStage[] = ['extraction', 'segmentation', 'reconstruction', 'tracking'];

export const STAGE_INFO: Record<PipelineStage, { name: string; description: string; number: number }> = {
  extraction: { name: 'Extraction', description: 'Extract frames from SVO2 files', number: 1 },
  segmentation: { name: 'Segmentation', description: 'Run SAM3 object detection', number: 2 },
  reconstruction: { name: '3D Reconstruction', description: 'Generate 3D bounding boxes', number: 3 },
  tracking: { name: 'Tracking', description: 'Track objects across frames', number: 4 },
};

export type StageStatus = 'pending' | 'running' | 'completed' | 'skipped';

export interface StageETA {
  stage: string;
  stage_number: number;
  status: StageStatus;
  eta_seconds: number | null;
  elapsed_seconds: number | null;
}

export interface JobConfig {
  object_class_ids?: string[];
  sam3_model_variant?: string;
  /** Confidence threshold for SAM3 detections (0-1) */
  sam3_confidence_threshold?: number;
  /** IOU threshold for non-max suppression */
  sam3_iou_threshold?: number;
  /** Batch size for SAM3 inference */
  sam3_batch_size?: number;
  /** Frame skip interval (1 = process all frames) */
  frame_skip?: number;
  /** Enable object tracking across frames */
  enable_tracking?: boolean;
  /** Export 3D bounding box data */
  export_3d_data?: boolean;
  /** Pipeline stages to run */
  stages_to_run?: PipelineStage[];
  /** Enable frame diversity filtering during extraction */
  enable_diversity_filter?: boolean;
  /** Similarity threshold for diversity filter (0-1, higher = more strict) */
  diversity_similarity_threshold?: number;
  /** Motion threshold for diversity filter (0-1, min motion to keep frame) */
  diversity_motion_threshold?: number;
}

export interface Job {
  id: string;
  name: string;
  status: JobStatus;
  current_stage?: number;
  current_stage_name?: string | null;
  progress?: number;
  stage_progress?: number;
  total_frames?: number | null;
  processed_frames?: number | null;
  total_detections?: number | null;
  input_paths?: string[];
  input_files?: string[];
  object_classes?: string[];
  output_directory?: string | null;
  // Storage tracking
  storage_size_bytes?: number | null;
  storage_size_formatted?: string | null;
  config?: JobConfig;
  stages_to_run?: PipelineStage[];
  dataset_id?: string | null;
  error_message?: string | null;
  created_at: string;
  started_at?: string | null;
  completed_at?: string | null;
  // ETA fields
  eta_seconds?: number | null;
  stage_etas?: StageETA[];
  frames_per_second?: number | null;
}

export interface JobListResponse {
  jobs: Job[];
  total: number;
  limit: number;
  offset: number;
}

export interface JobCreate {
  name: string;
  input_paths?: string[];
  input_files?: string[];
  object_classes?: string[];
  output_directory?: string;
  config?: Partial<JobConfig>;
}

export interface JobStatusUpdate {
  id: string;
  status: string;
  message: string | null;
}

export interface JobStatistics {
  total_frames: number;
  total_detections: number;
  detections_by_class: Record<string, number>;
  total_tracks: number;
  processing_time_seconds: number;
}

export interface JobResults {
  job_id: string;
  status: string;
  statistics: JobStatistics;
  output_directory: string;
  available_exports: string[];
}

// Alias for backward compatibility
export type JobResponse = Job;

// Pre-job duration estimation types

export interface StageEstimate {
  frames: number;
  estimated_seconds: number;
  fps: number;
}

export type EstimateConfidence = 'low' | 'medium' | 'high';

export interface EstimateDurationRequest {
  svo2_files: string[];
  frame_skip?: number;
  sam3_model_variant?: string;
  stages_to_run?: PipelineStage[];
  total_frames?: number;
}

export interface EstimateDurationResponse {
  estimated_total_frames: number;
  estimated_duration_seconds: number;
  estimated_duration_formatted: string;
  breakdown: Record<string, StageEstimate>;
  confidence: EstimateConfidence;
  based_on_jobs: number;
}

export interface FrameCountResponse {
  total_frames: number;
  files: Array<{
    path: string;
    frame_count: number | null;
    error: string | null;
  }>;
}

// Storage estimation types

export interface EstimateStorageRequest {
  total_frames: number;
  stages_to_run?: PipelineStage[];
  frame_skip?: number;
  extract_point_clouds?: boolean;
  extract_right_image?: boolean;
  extract_masks?: boolean;
  image_format?: string;
}

export interface EstimateStorageResponse {
  estimated_bytes: number;
  estimated_formatted: string;
  available_bytes: number;
  available_formatted: string;
  sufficient_space: boolean;
  warning: string | null;
  details: {
    frames_to_process: number;
    stages: string[];
    extract_point_clouds: boolean;
    extract_right_image: boolean;
    extract_masks: boolean;
  };
}

File: backend/app/schemas/job.py
"""Job-related Pydantic schemas."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from backend.app.constants import DEFAULT_PIPELINE_STAGES


class StageETA(BaseModel):
    """ETA information for a pipeline stage."""

    stage: str
    stage_number: int
    status: Literal["pending", "running", "completed", "skipped"]
    eta_seconds: int | None = None
    elapsed_seconds: int | None = None


class JobConfig(BaseModel):
    """Job configuration settings."""

    object_class_ids: list[str] = Field(description="IDs of object classes to detect")
    sam3_model_variant: str = "sam3_hiera_large"
    sam3_confidence_threshold: float = Field(default=0.5, ge=0.0, le=1.0)
    sam3_iou_threshold: float = Field(default=0.7, ge=0.0, le=1.0)
    sam3_batch_size: int = Field(default=8, ge=1, le=32)
    frame_skip: int = Field(default=1, ge=1, description="Process every Nth frame")
    enable_tracking: bool = True
    export_3d_data: bool = True
    stages_to_run: list[str] = Field(
        default_factory=lambda: DEFAULT_PIPELINE_STAGES.copy(),
        description="Pipeline stages to execute"
    )
    # Frame diversity filtering
    enable_diversity_filter: bool = Field(
        default=False,
        description="Filter out similar/redundant frames during extraction"
    )
    diversity_similarity_threshold: float = Field(
        default=0.85,
        ge=0.0,
        le=1.0,
        description="Threshold for visual similarity (higher = more strict)"
    )
    diversity_motion_threshold: float = Field(
        default=0.02,
        ge=0.0,
        le=1.0,
        description="Minimum motion required to keep a frame"
    )


class JobCreate(BaseModel):
    """Request schema for creating a new job."""

    name: str = Field(min_length=1, max_length=255)
    input_paths: list[str] = Field(default_factory=list, description="Paths to SVO2 files")
    output_directory: str | None = None
    config: JobConfig
    dataset_id: str | None = Field(
        default=None, description="Optional dataset ID to link job to dataset"
    )


class JobResponse(BaseModel):
    """Response schema for job details."""

    id: UUID
    name: str
    status: Literal["pending", "running", "paused", "completed", "failed", "cancelled"]
    current_stage: int = 0
    current_stage_name: str | None = None
    progress: float = 0.0
    stage_progress: float = 0.0
    total_frames: int | None = None
    processed_frames: int | None = None
    total_detections: int | None = None
    input_paths: list[str]
    output_directory: str | None = None
    storage_size_bytes: int | None = None
    storage_size_formatted: str | None = None
    config: JobConfig
    stages_to_run: list[str] = Field(
        default_factory=lambda: DEFAULT_PIPELINE_STAGES.copy()
    )
    dataset_id: UUID | None = None
    error_message: str | None = None
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    # ETA fields
    eta_seconds: int | None = None
    stage_etas: list[StageETA] = Field(default_factory=list)
    frames_per_second: float | None = None


class JobListResponse(BaseModel):
    """Response schema for job list."""

    jobs: list[JobResponse]
    total: int
    limit: int
    offset: int


class JobStatusUpdate(BaseModel):
    """Response schema for job status updates."""

    id: UUID
    status: str
    message: str | None = None


class JobStatistics(BaseModel):
    """Job processing statistics."""

    total_frames: int
    total_detections: int
    detections_by_class: dict[str, int]
    total_tracks: int
    processing_time_seconds: float


class JobResultsResponse(BaseModel):
    """Response schema for job results."""

    job_id: UUID
    status: str
    statistics: JobStatistics
    output_directory: str
    available_exports: list[str]


# Pre-job duration estimation schemas


class StageEstimate(BaseModel):
    """Estimate for a single pipeline stage."""

    frames: int = Field(description="Number of frames to process")
    estimated_seconds: int = Field(description="Estimated duration in seconds")
    fps: float = Field(description="Expected frames per second")


class EstimateDurationRequest(BaseModel):
    """Request schema for estimating job duration before starting."""

    svo2_files: list[str] = Field(
        description="Paths to SVO2 files (used to calculate total frames)"
    )
    frame_skip: int = Field(default=1, ge=1, description="Frame skip setting")
    sam3_model_variant: str = Field(
        default="sam3_hiera_large",
        description="SAM3 model variant affects segmentation speed"
    )
    stages_to_run: list[str] = Field(
        default_factory=lambda: ["extraction", "segmentation", "reconstruction", "tracking"],
        description="Pipeline stages to estimate"
    )
    # Optional: if total_frames already known, skip frame counting
    total_frames: int | None = Field(
        default=None,
        description="If provided, skip frame counting from files"
    )


class EstimateDurationResponse(BaseModel):
    """Response schema for job duration estimate."""

    estimated_total_frames: int = Field(description="Total frames to process after frame_skip")
    estimated_duration_seconds: int = Field(description="Total estimated duration")
    estimated_duration_formatted: str = Field(
        description="Human-readable duration (e.g., '~7 min')"
    )
    breakdown: dict[str, StageEstimate] = Field(
        description="Per-stage duration breakdown"
    )
    confidence: Literal["low", "medium", "high"] = Field(
        description="Estimate confidence based on historical data"
    )
    based_on_jobs: int = Field(
        description="Number of historical jobs used for benchmark"
    )


# Storage estimation schemas


class EstimateStorageRequest(BaseModel):
    """Request schema for estimating job storage before creation."""

    total_frames: int = Field(description="Total frames in input files")
    stages_to_run: list[str] = Field(
        default_factory=lambda: ["extraction", "segmentation", "reconstruction", "tracking"],
        description="Pipeline stages to run"
    )
    frame_skip: int = Field(default=1, ge=1, description="Frame skip setting")
    extract_point_clouds: bool = Field(default=True, description="Whether to extract point clouds")
    extract_right_image: bool = Field(default=True, description="Whether to extract right camera image")
    extract_masks: bool = Field(default=True, description="Whether to save segmentation masks")
    image_format: str = Field(default="png", description="Image format (png or jpg)")


class EstimateStorageResponse(BaseModel):
    """Response schema for job storage estimate."""

    estimated_bytes: int = Field(description="Estimated storage in bytes")
    estimated_formatted: str = Field(description="Human-readable size (e.g., '25.5 GB')")
    available_bytes: int = Field(description="Available disk space in bytes")
    available_formatted: str = Field(description="Human-readable available space")
    sufficient_space: bool = Field(description="Whether there's enough space for the job")
    warning: str | None = Field(default=None, description="Warning message if space is low")
    details: dict = Field(description="Breakdown of estimation parameters")
